### PR TITLE
H-1466: Revisit parameters in `Store` functions

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

The amount of parameters in the `*Store` traits gets overwhelming and the Interface of the actual API, the OpenAPI, and the integration tests diverge a fair bit.

## 🔍 What does this change?

- Revisits large parts of the API to use structs as arguments instead of plain parameters to allow named arguments
- Re-use many of the structs in the OpenAPI so the Node API uses a more similar representation of the Graph API

### 🚢 Has this modified a publishable library?

This PR:

- [x] does not modify any publishable blocks or libraries, or modifications do not need publishing

### 📜 Does this require a change to the docs?

The changes in this PR:

- [x] are internal and do not require a docs change

### 🕸️ Does this require a change to the Turbo Graph?

The changes in this PR:

- [x] do not affect the execution graph

## 🐾 Next steps

- ~So far, only data types, accounts, and entities were adjusted. property types and entity types has to be implemented before this PR should be merged.~
- Add a unified batch endpoint where the structs can be passed as a list

### CODE REVIEW

File: apps/hash-graph/libs/api/src/rest/data_type.rs [hunk 1]
```diff
@@ -541,24 +531,22 @@ struct UpdateDataTypeEmbeddingsRequest {
...
-    request_body = UpdateDataTypeEmbeddingsRequest,
+    request_body = UpdateDataTypeEmbeddingParams,
 )]
 #[tracing::instrument(level = "info", skip(store_pool, authorization_api_pool))]
 async fn update_data_type_embeddings<S, A>(
     AuthenticatedUserHeader(actor_id): AuthenticatedUserHeader,
     store_pool: Extension<Arc<S>>,
     authorization_api_pool: Extension<Arc<A>>,
-    body: Json<UpdateDataTypeEmbeddingsRequest>,
+    Json(body): Json<serde_json::Value>,
```
Comment (indietyp) [inline apps/hash-graph/libs/api/src/rest/data_type.rs:541]: Maybe write a comment on why this is needed. It might not be obvious to the uninitiated ones.

File: apps/hash-graph/libs/api/src/rest/data_type.rs [hunk 2]
```diff
@@ -203,31 +205,25 @@ where
...
             actor_id,
             &mut authorization_api,
             temporal_client.as_deref(),
-            data_types.into_iter().zip(partial_metadata),
-            ConflictBehavior::Fail,
-            relationships,
+            schema_iter.into_iter().map(|schema| {
+                domain_validator.validate(&schema).map_err(|report| {
+                    tracing::error!(error=?report, id=schema.id().to_string(), "Data Type ID failed to validate");
+                    StatusCode::UNPROCESSABLE_ENTITY
```
Comment (indietyp) [inline apps/hash-graph/libs/api/src/rest/data_type.rs:217]: down there changed to `hash_status`, maybe here as well?

File: apps/hash-graph/libs/api/src/rest/web.rs [hunk 3]
```diff
@@ -114,14 +107,12 @@ async fn create_web<S, A>(
...
 ) -> Result<StatusCode, StatusCode>
 where
     S: StorePool + Send + Sync,
     A: AuthorizationApiPool + Send + Sync,
 {
-    let CreateWebRequest { owned_by_id, owner } = body;
-
     let mut store = store_pool.acquire().await.map_err(|report| {
         tracing::error!(error=?report, "Could not acquire store");
         StatusCode::INTERNAL_SERVER_ERROR
```
Comment (indietyp) [inline apps/hash-graph/libs/api/src/rest/web.rs:118]: change to `hash_status`

File: apps/hash-graph/libs/graph/src/store/fetcher.rs [hunk 4]
```diff
@@ -510,9 +522,15 @@ where
...
-                    ConflictBehavior::Skip,
-                    DATA_TYPE_RELATIONSHIPS,
+                    fetched_ontology_types
+                        .data_types
+                        .into_iter()
+                        .map(|(data_type, metadata)| CreateDataTypeParams {
+                            schema: data_type,
+                            classification: metadata.classification,
+                            relationships: DATA_TYPE_RELATIONSHIPS,
+                            conflict_behavior: ConflictBehavior::Fail,
```
Comment (indietyp) [inline apps/hash-graph/libs/graph/src/store/fetcher.rs:532]: why the change from `Skip` to `Fail`?

File: apps/hash-graph/libs/graph/src/store/fetcher.rs [hunk 5]
```diff
@@ -1237,44 +1238,20 @@ where
...
-                properties,
-                link_data,
-                relationships,
-            )
+            .create_entity(actor_id, authorization_api, temporal_client, params)
             .await
     }
 
-    #[tracing::instrument(level = "info", skip(self, authorization_api))]
+    #[tracing::instrument(level = "info", skip(self, authorization_api, params))]
```
Comment (indietyp) [inline apps/hash-graph/libs/graph/src/store/fetcher.rs:1245]: Why? I feel like we're omitting useful additional information here.

**Review by indietyp — CHANGES_REQUESTED** [review]
partial review. A lot to go through (rn 23/38 done)

File: apps/hash-graph/libs/api/src/rest/data_type.rs (hunk 2 above)
Comment (TimDiekmann) [reply apps/hash-graph/libs/api/src/rest/data_type.rs:217]: The return value of the Function is of type `http::StatusCode`, thus this must not be a `hash_status`. As we're going to switch to our own RPC implementation soon(-ish) I'm only going to change the signatures which are required in this PR.

File: apps/hash-graph/libs/api/src/rest/web.rs (hunk 3 above)
Comment (TimDiekmann) [reply apps/hash-graph/libs/api/src/rest/web.rs:118]: Same as above

File: apps/hash-graph/libs/graph/src/store/fetcher.rs (hunk 4 above)
Comment (TimDiekmann) [reply apps/hash-graph/libs/graph/src/store/fetcher.rs:532]: Did this by accident, thanks for spotting!

File: apps/hash-graph/libs/graph/src/store/fetcher.rs (hunk 5 above)
Comment (TimDiekmann) [reply apps/hash-graph/libs/graph/src/store/fetcher.rs:1245]: Actually, I removed this instrumentation as this is not really relevant. It's just calling the underlying function which has an instrumentation as well. I removed `params` there.

**Comment by TimDiekmann** [comment]
Thanks, Bilal, very helpful review so far! I applied your requested changes.

File: apps/hash-graph/libs/api/src/rest/data_type.rs (hunk 1 above)
Comment (TimDiekmann) [reply apps/hash-graph/libs/api/src/rest/data_type.rs:543]: I added a comment where the deserialization happens

File: apps/hash-graph/libs/api/src/rest/account.rs [hunk 6]
```diff
@@ -171,9 +180,9 @@ where
         StatusCode::INTERNAL_SERVER_ERROR
     })?;
 
-    let account_group_id = AccountGroupId::new(Uuid::new_v4());
+    let account_group_id = params.account_group_id;
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply apps/hash-graph/libs/api/src/rest/account.rs:183]: No. `params.account_group_id` is defined as 
```
fn random_account_group_id() -> AccountGroupId {
    AccountGroupId::new(uuid::Uuid::new_v4())
}

#[derive(Deserialize)]
pub struct InsertAccountGroupIdParams {
    #[serde(default = "random_account_group_id")]
    pub account_group_id: AccountGroupId,
}
```

So it's now possible to specify an account (group) id, but by default a random one is generated.

File: apps/hash-graph/libs/graph/src/store/postgres/knowledge/entity/mod.rs [hunk 7]
```diff
@@ -484,15 +483,15 @@ impl<C: AsClient> EntityStore for PostgresStore<C> {
         let (edition_id, closed_schema) = transaction
             .insert_entity_edition(
                 edition_created_by_id,
-                archived,
-                &entity_type_url,
-                &properties,
+                false,
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply apps/hash-graph/libs/graph/src/store/postgres/knowledge/entity/mod.rs:486]: It should not be possible to create an archived entity (note, that we do not support non-link entities and entity archival is not properly implemented through temporal versioning).

File: apps/hash-graph/libs/graph/src/store/postgres/knowledge/entity/mod.rs [hunk 8]
```diff
@@ -1284,31 +1259,49 @@ impl<C: AsClient> EntityStore for PostgresStore<C> {
...
                         DELETE FROM entity_embeddings
-                        WHERE (web_id, entity_uuid) IN (
-                            SELECT *
-                            FROM UNNEST($1::UUID[], $2::UUID[])
-                        )
-                        AND updated_at_transaction_time <= $3
-                        AND updated_at_decision_time <= $4;
+                        WHERE web_id = $1
+                          AND entity_uuid = $2
+                          AND draft_id = $3
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply apps/hash-graph/libs/graph/src/store/postgres/knowledge/entity/mod.rs:1270]: It was forgotten previously when I implemented draft IDs.

File: apps/hash-graph/libs/graph/src/store/fetcher.rs [hunk 9]
```diff
@@ -586,9 +637,15 @@ where
...
-                        ConflictBehavior::Skip,
-                        DATA_TYPE_RELATIONSHIPS,
+                        fetched_ontology_types
+                            .data_types
+                            .into_iter()
+                            .map(|(schema, metadata)| CreateDataTypeParams {
+                                schema,
+                                classification: metadata.classification,
+                                relationships: DATA_TYPE_RELATIONSHIPS,
+                                conflict_behavior: on_conflict,
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply apps/hash-graph/libs/graph/src/store/fetcher.rs]: Thanks: c33ba304a28e12a1e878f82c24bacbc734883434

File: apps/hash-graph/libs/graph/src/store/fetcher.rs [hunk 10]
```diff
@@ -601,9 +658,14 @@ where
...
                         temporal_client,
-                        fetched_ontology_types.property_types,
-                        ConflictBehavior::Skip,
-                        PROPERTY_TYPE_RELATIONSHIPS,
+                        fetched_ontology_types.property_types.into_iter().map(
+                            |(schema, metadata)| CreatePropertyTypeParams {
+                                schema,
+                                classification: metadata.classification,
+                                relationships: PROPERTY_TYPE_RELATIONSHIPS,
+                                conflict_behavior: on_conflict,
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply apps/hash-graph/libs/graph/src/store/fetcher.rs]: Thanks: c33ba304a28e12a1e878f82c24bacbc734883434

File: apps/hash-graph/libs/graph/src/store/fetcher.rs [hunk 11]
```diff
@@ -616,9 +678,16 @@ where
...
-                        ConflictBehavior::Skip,
-                        ENTITY_TYPE_RELATIONSHIPS,
+                        fetched_ontology_types.entity_types.into_iter().map(
+                            |(schema, metadata)| CreateEntityTypeParams {
+                                schema,
+                                classification: metadata.classification,
+                                icon: metadata.icon,
+                                label_property: metadata.label_property,
+                                relationships: ENTITY_TYPE_RELATIONSHIPS,
+                                conflict_behavior: on_conflict,
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply apps/hash-graph/libs/graph/src/store/fetcher.rs]: Thanks: c33ba304a28e12a1e878f82c24bacbc734883434

**Review by indietyp — APPROVED** [review]
